src/mlir/lowering.md

```diff
@@ -0,0 +1,517 @@
+a similar struct representation, since every field read/write should have the
+class type and the field name, with the concrete field type as a result.
+
+## Regions
+
+Every object is allocated into a region. Some objects share regions, others
+live in immutable memory. Whether we'll allow users to annotate objects with
+region information (ex. `new Foo in Bar`) or not is up to debate. Regardless,
+we'll need that kind of information in the IR to add calls to the right runtime
+library functions that control memory.
+
+We want a pass that bundles objects into regions based on their access patterns
+taking user annotations as either hints or requirements. Such pass could be done
+in 1IR or 2IR, but it will depending where it runs and whether we allow user
+annotations will dictate if the 1IR will have region information, but the 2IR
+will definitely need it.
+
+Region informaiton can be represented as references to the owning `iso` (see
+below) on the types on every instance of, to avoid repeated and potentially
+complex search patterns.
+
+### ISO
+
+Iso references are region owning pointers that can freely access the whole region
+and sub-regions (if not given to another task). Any read-write access through
+those references is free.
+
+### MUT
+
+Mutable references must belong to a region and must have access to that region,
+ie. they need to belong to the owning `iso` of the region. For this, mutable
+references must point to the `iso` references. This can be represented as
+MLIR annotation, referring the SSA value (ex. `#region<%22>` where `%22` must be
+an `iso` reference of the same region).
+
+### drop
+
+Dropping (deallocating) regions need to know that there are no more active users
+of that region (ex. reference counting) and also need to find which region it is.
+If we're dropping an `iso`, it's straight forward, if it's a `mut`, we simply
+follow the annotation to find the `iso`.
+
+The actual deallocation will be a call to the runtime library if the region is
+gone (ie. equals `nullptr`). This is a simple check that can be lowered by the
+translation directly.
+
+## Statements
+
+### NewStmt
+
+Allocates a new region or more memory in an existing region. We should keep
+this as a dialect operation until we do region optimisation. This is not the
+same as region tracking in 1IR, which will validate regions and insert `merge`
+nodes where necessary.
+
+The `NewStmt` will be lowered as `verona.new` and merge nodes as `verona.merge`
+and both will allow us to optimise region merges using flow analysis (and other
+tricks) to merge regions earlier on (replacing `merge(a, b)` with `new b in a`,
+for example).
+
+The operation has two semantics plus the merge:
+```
+  // new object in a new region
+  %0 = verona.new_region @Class [initialiser]
+
+  // new object in an existing region
+  %1 = verona.new_object @Class [initialiser] in @Other
+
+  // Merge two regions
+  verona.merge %0, %1
```

Comment:
`merge` returns a mut reference to the object `%1` referenced (`%1` would have been an `iso`).